heatmap/src/heatmap.rs

```diff
@@ -2,34 +2,41 @@
 // Licensed under the Apache License, Version 2.0
 // http://www.apache.org/licenses/LICENSE-2.0
 
-use crate::window::OwnedWindow;
 use crate::Error;
 use crate::*;
 use core::sync::atomic::*;
 
+use clocksource::DateTime;
 use histogram::{Bucket, Histogram};
 use parking_lot::Mutex;
 
 /// A `Heatmap` stores counts for timestamped values over a configured span of
 /// time.
 ///
 /// Internally, it is represented as a ring buffer of histograms with one
-/// additional histogram to track all counts within the span of time. Old
-/// histograms age-out as time moves forward and they are subtracted from the
+/// additional histogram ("summary") to track all counts within the span of time.
+/// Each histogram covers a time slice whose width is specified by `resolution`.
+/// Old histograms age-out as time moves forward and they are subtracted from the
 /// summary histogram at that point.
 ///
 /// This acts as a moving histogram, such that requesting a percentile returns
 /// a percentile from across the configured span of time.
 pub struct Heatmap {
-    slices: Vec<OwnedWindow>,
-    current: AtomicUsize,
-    lock: Mutex<()>,
-    start: AtomicInstant,
-    stop: AtomicInstant,
     span: Duration,
     resolution: Duration,
-    next_tick: AtomicInstant,
+
     summary: Histogram,
+    histograms: Vec<Histogram>,
+
+    // index in the ring buffer of the current histogram, which has the latest start time
+    idx: AtomicUsize,
+    // start time of the current time slice
+    create_at: DateTime,
```

Comment:
I don't see this being updated? Am I missing something?

Reply:
line 153

Reply:
I think my confusion is that `create_at` seems less useful once the heatmap has begun to clear and re-use slices. I don't see a way in the current implementation to figure out what the `DateTime` is for the oldest slice once that happens.

Reply:
I've reworked the code to calculate oldest slice and whether it's looped around without using this field.